src/internal/packager/git/push.go

```diff
@@ -18,6 +18,10 @@ import (
 )
 
 func (g *Git) PushRepo(localPath string) error {
+	// Keeep a copy of the original address in case we need to retry this push
+	originalAddress := g.Server.Address
+	defer g.setServerAddress(originalAddress)
```

Comment:
Style thoughts on doing the following block instead? (so we don't need to make a setter func)
```
defer func() { g.Server.Address = originalAddress }()
```

Reply:
I think what's currently in the PR is more clear, but I generally dislike getter/setter functions so I could go either way.

Reply:
I prefer `defer func() { g.Server.Address = originalAddress }()` but also come from a dirty JavaScript background where that kind of thing is normal.  I like setters when they are going to be reused, but not when they aren't.

Reply:
I recommend we not do this at all and instead just recreate the variable in the outer loop to keep things clean. Defers are not guaranteed to run and this looks like a recipe for future bugginess.  I would love to rewrite the caller to something like this below (untested), thoughts?

<img width="844" alt="Screenshot 2022-12-02 at 2 38 08 PM" src="https://user-images.githubusercontent.com/882485/205381650-1dbd6cc4-8597-43c3-8fe0-36578ef3d2af.png">

```golang

// Push all of the components git repos to the configured git server
func (p *Packager) pushReposToRepository(reposPath string, repos []string) error {
	// Try repo push up to 3 times
	for _, repoURL := range repos {
		// Reusable function to attempt to push a git repo
		tryPush := func() error {
			gitClient := git.New(p.cfg.State.GitServer)

			// Convert the repo URL to a Zarf-formatted repo name
			if repoPath, err := gitClient.TransformURLtoRepoName(repoURL); err != nil {
				return fmt.Errorf("unable to get the repo name from the URL %s: %w", repoURL, err)
			} else {
				// On successful conversion, try to push the repo
				return gitClient.PushRepo(filepath.Join(reposPath, repoPath))
			}
		}

		// Try to push the repo up to 3 times
		if err := utils.Retry(tryPush, 3, 5*time.Second); err != nil {
			return fmt.Errorf("unable to push repo %s to the Git Server: %w", repoURL, err)
		}
	}

	return nil
}
```

Reply:
Looking through the recommendation now, but

> Defers are not guaranteed to run

That is kind of news to me. I think the only time I know defers don't run is if an `os.Exit()` is called which would make this fix mute anyways.

Reply:
This implementation would have the function `(g *Git) PushRepo(...)` modifying the `g.Server.Address` to an IP/PORT combo that no longer is accessible after the function returns. That is fine with this specific implementation since we are throwing away the `gitClient` right after this push is completed but that is not an intuitive thing and people using that function could easily slip up.

@jeff-mccoy What do you think about turning `(g *Git) PushRepo(...)` into a more pure wrapper around `git push` and force the caller to create the k8s tunneled address before calling?

Reply:
^ Above comment is assuming we remove the `defer g.setServerAddress()` like you were implying we should

Reply:
I would prefer that we purify the push / pull functions into more pure drivers of git (like clone does) and then we can call those functions in a retry loop more easily from a driver function (that way we still have one client and that client wont keep getting recalled in the retry loop.  IMO that provides better separation of concerns and won't unnecessarily create more and more clients in memory.

Reply:
Don't know if I'm tracking the clients in memory comment, yes new would be crated and GCd, but they are very small, it's not giant like packager.

Reply:
Yeah its not a big memory leak for sure, but I just feel it is generally cleaner to separate the the overall driving function, the receiver client and the wrapper around git (and then the retry logic can more happily live inside of the receiver instead of being mixed in with git operations).  Also I usually just prefer to not rely on GC if you don't have to